### ISSUE TITLE
feat: CodeQualityScanner — high-risk coding pattern detection

### ISSUE THREAD

**Issue by floswald** [floswald/PackageScanner.jl#4]
## Summary

Extend PackageScanner with a new module `CodeQualityScanner.jl` that detects high-risk coding patterns in replication packages. This complements PII scanning with a code quality / reproducibility risk layer. Findings are reported as a dedicated section in the existing report output.

---

### Architecture

- New file `src/CodeQualityScanner.jl` as a module inside PackageScanner.jl
- Pattern definitions in `config/code_quality_patterns.toml`, keyed by language and rule name, so rules can be extended without touching Julia code
- Each rule has: `pattern` (regex), `risk_level` (`:critical` or `:advisory`), `message` (human-readable explanation for the report)
- Two-pass logic where needed (seed check): first collect evidence of stochastic calls, then check for seed presence across the whole package

### Existing infrastructure

* existing function `classify_files` classififies all files based on file endings into either code, data, or docs. use the output of this function.
* new functionality needs to be integrated into function `precheck_package` after line 105 where it says currently

```
    @info "Parse code files and search for filepaths"
    file_paths(codefiles, out)
```

here we need a call to the to be implemented function on `codefiles`.

---

### Rule 1: Missing random seed (all languages)

The check is package-level, not line-level:
1. Scan all files of a given language for any stochastic function call (triggers)
2. If triggers found, scan all files of that language for any seed-setting call
3. Flag as **critical** if triggers exist but no seed call found anywhere

**Stata**

Triggers: `bootstrap`, `bsample`, `simulate`, `drawnorm`, `permute`, `randorder`
Seed calls: `set seed`

**R**

Triggers: `rnorm`, `runif`, `rbinom`, `sample(`, `replicate`, `boot(`, any `_cv(` or `trainControl`
Seed calls: `set.seed(`

**Python**

Triggers: `np.random`, `random.`, `sklearn`, `train_test_split`, `shuffle(`
Seed calls: `random.seed(`, `np.random.seed(`, `random_state=`, `torch.manual_seed`, `tf.random.set_seed`

**Julia**

Triggers: `rand(`, `randn(`, `shuffle(`, `sample(`
Seed calls: `Random.seed!`, `seed!(`

---

### Rule 2: Hardcoded absolute paths (all languages)

Single-pass regex per file, flag each occurrence.

- **Stata:** paths matching `["']\/Users\/`, `["']C:\\`, `["']\/home\/`
- **R / Python / Julia:** same patterns, plus `os.path`, `Path(` containing a home directory prefix

Risk: **critical** — package will not run on another machine.

---

### Rule 3: `merge m:m` (Stata only)

Flag every occurrence. Risk: **critical**.

Message: *"`merge m:m` uses positional row-matching within key groups, not relational join semantics. Use `joinby` for a true many-to-many join, or identify the correct unique key and use `1:m`/`m:1`."*

---

### Rule 4: Undocumented sample drops (Stata, R, Python)

Flag `drop if` / `keep if` (Stata), `filter(` (R), `.query(` / `.loc[` (Python) that are not preceded by a comment within 2 lines.

Risk: **advisory** — not necessarily wrong, but warrants author explanation.

---

### Rule 5: Merge without explicit join type (R, Python)

- **R:** `merge(` calls missing `all=`, `all.x=`, `all.y=` arguments
- **Python:** `pd.merge(` or `.merge(` calls missing `how=` argument

Risk: **advisory** — silently defaults to inner join, may drop rows unexpectedly.

---

### Report integration

Add a `## Code Quality` section to the existing report, grouped by language, then by risk level. Example:

```
## Code Quality

### Stata
[CRITICAL] merge m:m detected (main_analysis.do, line 42)
  → merge m:m uses positional row-matching...

[CRITICAL] No random seed set — stochastic calls detected (bootstrap on line 88)

### R
[ADVISORY] merge() called without explicit all= argument (build_panel.R, line 14)
```

---

### Config file sketch

```toml
[stata.merge_mm]
pattern = "merge m:m"
risk = "critical"
message = "merge m:m uses positional row-matching..."

[stata.seed_trigger]
patterns = ["bootstrap", "bsample", "simulate", "drawnorm"]

[stata.seed_call]
patterns = ["set seed"]
```

---

## Out of scope (for now)

- Control flow analysis to verify seed is set *before* stochastic call (requires execution DAG)
- Cross-script seed propagation checking
- R `merge m:m` equivalent patterns (flagged separately under Rule 5)

---

## Acceptance criteria

- [ ] `CodeQualityScanner.jl` module exists and is callable from PackageScanner
- [ ] TOML config drives all pattern definitions
- [ ] Seed check is package-level (not line-level) for all four languages
- [ ] Report output includes a dedicated `## Code Quality` section
- [ ] Rules 1–3 implemented and tested; Rules 4–5 implemented as best-effort regex
